rebuild.sh

```diff
@@ -122,6 +122,14 @@ mvnBuildDocker() {
     *)
       mvnImage=maven:${mvnVersion}-eclipse-temurin-${jdk}-alpine
   esac
+  if ! docker pull -q ${mvnImage} > /dev/null 2>&1; then
```

Comment:
@Stephan202 I tested this (with bin/update-statistics.sh) and found that `docker pull -q` fails when the image is already available

Reply:
Interesting; locally I can't reproduce this:
```sh
$ docker --version
Docker version 20.10.20, build 9fdeb9c
$ docker pull -q maven:3.8.5-eclipse-temurin-17-alpine 2>&1 2>/dev/null; echo $?
docker.io/library/maven:3.8.5-eclipse-temurin-17-alpine
0
$ docker pull -q maven:3.8.5-eclipse-temurin-17-alpine 2>&1 2>/dev/null; echo $?
docker.io/library/maven:3.8.5-eclipse-temurin-17-alpine
0
$ docker pull -q maven:this-one-does-not-exist 2>&1 2>/dev/null; echo $?
1
```

Perhaps it depends on the version of Docker used (which one do you have installed locally?).

I can have a closer look at `bin/update-statistics.sh` later; right now I'm commuting; don't want to pull new images while using mobile internet :)

Reply:
> I can have a closer look at `bin/update-statistics.sh` later.

Didn't happen yesterday, and today is full of meeting and events; I intend to circle back to this PR later this week. Sorry for the delay.

Reply:
no worries, we all have a day job, I have the same type of delays that you patiently waited for

Reply:
I just tested (and improved a little bit the script): you're right, `docker pull -q` works, I probably was hit by my rate limit or anything stupid and transient

now, with the `bin/update-statistics.sh` script, we can easily evaluate the impact of any change in image tag selection and see if it breaks any buildspec tool x jdk selection

I let you review and update the statistics script while changing `rebuild.sh` so we can merge with safety

Reply:
Finally found a moment to get back to this :)

I saw that `bin/update-statistics.sh` uses `docker inspect` rather than `docker manifest inspect`. So this means there are three options. I wrote the following script to compare them:
```sh
#!/usr/bin/env bash

# The image against which to test.
image=maven:3.6.3-openjdk-17-slim

print_header() {
  local header="${1}"
  echo -e "\n\033[0;34m${header}\033[0m"
}

print_result() {
  test $? == 0 && echo -e '\033[0;32mSUCCESS\033[0m' || echo -e '\033[0;31mFAILURE\033[0m'
}

# Drop the image if it exists.
docker image rm -f "${image}"

print_header 'Running `docker inspect` while the image is locally absent.'
time docker inspect "${image}" > /dev/null
print_result

print_header 'Running `docker manifest inspect` while the image is locally absent.'
time docker manifest inspect "${image}" > /dev/null
print_result

print_header 'Running `docker pull -q` while the image is locally absent.'
time docker pull -q "${image}" > /dev/null
print_result

print_header 'Running `docker inspect` while the image is locally present.'
time docker inspect "${image}" > /dev/null
print_result

print_header 'Running `docker manifest inspect` while the image is locally present.'
time docker manifest inspect "${image}" > /dev/null
print_result

print_header 'Running `docker pull -q` while the image is locally present.'
time docker pull -q "${image}" > /dev/null
print_result
```

Based on the results, it seems to me that:
1. `docker inspect` is by far the fastest, **but** only says whether the given image exists locally. It seems this command cannot be used in `rebuild.sh`, unless there's some kind of "pre-population" of images. (Which might be worth the effort of implementing; see below.)
2. `docker manifest inspect` takes ~3 seconds if the image doesn't exist, and ~5 if it does, _irrespective of how often the command is executed_.
3. `docker pull -q` takes ~2 seconds if the image doesn't exist, and ~1.5 if is does, _unless the image doesn't exist locally yet, in which case it of course takes longer_ (but that's a cost that needs to be paid regardless).

Based on this I see two options:
1. Stick with `docker pull -q` in `rebuild.sh`.
2. Introduce a `docker pull -q`-based "bootstrap phase" and use `docker inspect` in `rebuild.sh`.

(The timings presented here _may_ of course be specific to my setup; interested to hear whether you can reproduce these results.)